README.md
# binlogik

binlogik is a ruby-based mysql binlog utility.

its default mode emits mysql binlog query and row events to stdout, and runs as a kubernetes pod, detecting `MYSQL_SERVICE` from the `ENV` via `superconfig2`

# related work / research links

https://dev.mysql.com/blog-archive/more-metadata-is-written-into-binary-log/

https://github.com/shyiko/mysql-binlog-connector-java/issues/24#issuecomment-446369124

https://dev.mysql.com/doc/dev/mysql-server/latest/classbinary__log_1_1Table__map__event.html

https://dev.mysql.com/doc/dev/mysql-server/latest/classbinary__log_1_1Table__map__event.html#Table_table_map_event_optional_metadata

https://github.com/mysql/mysql-server/commit/c019294ca8fece7af3bca6e6190e4e1efafa22af

https://github.com/mysql/mysql-server/blob/8.0/libbinlogevents/include/rows_event.h#L383

https://github.com/mysql/mysql-server/blob/8.0/libbinlogevents/include/rows_event.h#L50

https://dev.mysql.com/doc/refman/8.0/en/mysqlbinlog.html

https://dev.mysql.com/doc/refman/8.0/en/mysql-command-options.html

https://github.com/mysql/mysql-server/commit/c019294ca8fece7af3bca6e6190e4e1efafa22af

https://dev.mysql.com/doc/dev/mysql-server/latest/rows__event_8h_source.html

https://dev.mysql.com/doc/internals/en/query-event.html

https://dev.mysql.com/doc/internals/en/binlog-event.html

https://dev.mysql.com/doc/internals/en/com-binlog-dump.html

https://dev.mysql.com/doc/refman/5.7/en/purge-binary-logs.html

https://dev.mysql.com/doc/dev/mysql-server/8.0.21/classQuery__log__event.html

https://dev.mysql.com/doc/internals/en/xid-event.html

https://www.rubydoc.info/gems/mysql_binlog/0.3.2/MysqlBinlog

https://dev.mysql.com/doc/dev/mysql-server/latest/statement__events_8h_source.html

https://dev.mysql.com/doc/refman/5.7/en/mysqlbinlog.html#option_mysqlbinlog_start-position

https://github.com/y310/kodama

https://github.com/krowinski/php-mysql-replication/blob/41073226e6c35c793499ba928ae02d81f1a47b8a/src/MySQLReplication/BinLog/BinLogSocketConnect.php

https://github.com/jeremycole/mysql_binlog/blob/master/lib/mysql_binlog/binlog.rb

https://github.com/jeremycole/mysql_binlog/blob/master/lib/mysql_binlog/reader/binlog_stream_reader.rb

https://www.rubydoc.info/gems/mysql_binlog/0.2.0/MysqlBinlog/BinlogStreamReader

https://github.com/jeremycole/mysql_binlog/blob/ac425f6688b523f016725dab8af3a708655c8a68/lib/mysql_binlog/reader/binlog_stream_reader.rb

https://github.com/xiongbill/mysql_binlog/blob/master/bin/mysql_binlog_dump

https://www.rubydoc.info/gems/mysql_binlog/0.3.2/MysqlBinlog/BinlogStreamReader

https://ruby-doc.org/stdlib-3.0.0/libdoc/socket/rdoc/Socket.html

https://www.google.com/search?channel=fs&client=ubuntu&q=ruby+mysql2+COM_BINLOG_DUMP

https://github.com/tmtm/ruby-mysql/blob/44520a12fac454a41a3759256f42fea04f0032d3/lib/mysql/protocol.rb

https://dev.mysql.com/doc/internals/en/com-binlog-dump-gtid.html

https://dev.mysql.com/doc/internals/en/com-binlog-dump.html

https://stackoverflow.com/questions/66567796/how-can-i-send-commandslike-com-register-slave-or-com-binlog-dump-to-the-mysql

https://dev.mysql.com/doc/c-api/8.0/en/mysql-binlog-fetch.html

https://github.com/brianmario/mysql2/blob/master/lib/mysql2/client.rb

https://www.rubydoc.info/gems/mysql_binlog/0.2.0/MysqlBinlog/BinlogStreamReader

https://www.rubydoc.info/gems/ruby-mysql/2.9.14/Mysql

https://github.com/noplay/python-mysql-replication

https://github.com/SponsorPay/ruby-binlog/blob/master/ext/ruby_binlog_event.h

https://github.com/SponsorPay/ruby-binlog/blob/master/ext/ruby_binlog_event.h

https://github.com/jeremycole/mysql_binlog

https://github.com/y310/kodama

https://github.com/SponsorPay/ruby-binlog/blob/master/ext/ruby_binlog.cpp

https://github.com/zalora/binlog-parser

https://github.com/twingly/ecco

https://github.com/jeremycole/mysql_binlog/blob/master/lib/mysql_binlog/reader/binlog_file_reader.rb

http://shlomi-noach.github.io/awesome-mysql/

https://github.com/brianmario/mysql2

https://github.com/brianmario/mysql2/blob/a825ba14af8c6def7d17d7562cc554db22043bd3/spec/mysql2/statement_spec.rb

https://github.com/brianmario/mysql2/blob/926f85382b0e6e8b051531e77f373cfbcea7365c/README.md

https://github.com/brianmario/mysql2/blob/b439a895ef6b289e1bc5e07303fc3952713fb948/lib/mysql2.rb

https://github.com/brianmario/mysql2/blob/7f4e844fccf6afa888d0bd108d4707a2a7784484/ext/mysql2/client.c

https://github.com/brianmario/mysql2/issues/771

https://github.com/jeremycole/mysql_binlog/commit/1f473eb77eb4cce709510de7c7bee73ba006a007

https://github.com/google/mysql/blob/master/client/mysqlbinlog.cc

https://github.com/tmtm/ruby-mysql/issues/5

https://www.rubydoc.info/gems/ruby-mysql-ext/2.9.11/Mysql/Protocol

https://www.google.com/search?channel=fs&client=ubuntu&q=binlog_api.h+mysql

https://github.com/Flipkart/MySQL-replication-listener/blob/master/include/binlog_api.h

https://dev.mysql.com/doc/refman/5.7/en/mysqlbinlog.html#option_mysqlbinlog_start-datetime

http://manpages.ubuntu.com/manpages/focal/man1/mysqlbinlog.1.html

https://github.com/tmtm/ruby-mysql/blob/master/lib/mysql/protocol.rb

https://github.com/jeremycole/mysql_binlog/blob/master/lib/mysql_binlog/reader/binlog_file_reader.rb

https://github.com/krowinski/php-mysql-replication

https://github.com/shyiko/mysql-binlog-connector-java/blob/dd710a5466381faa57442977b24fceff56a0820e/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/EventDeserializer.java

https://github.com/shyiko/mysql-binlog-connector-java/blob/dd710a5466381faa57442977b24fceff56a0820e/src/main/java/com/github/shyiko/mysql/binlog/io/ByteArrayInputStream.java

https://github.com/jeremycole/mysql_binlog

https://dev.mysql.com/doc/refman/5.7/en/replication-options-binary-log.html

https://github.com/mavenlink/changestream/blob/8c362a6609524e01a8dd370e38441b86e07e0814/docs/binlog-event-routing.md

https://github.com/ruby/ruby/blob/5e7675d442d968825fec100b915905a7706c981f/io.c#L6426


CHANGELOG.md
# [2.3.0] - 2023-01-29 - Jon Bardin

add rudimentary support for binlog_row_metadata=FULL

#######

# [2.2.0] - 2022-01-29 - Jon Bardin

Add better support for update/insert row event parsing, simplify local debug case

#######

# [2.1.0] - 2021-09-05 - Jon Bardin

Add debug utility to detect lag between primary instance and replica nodes detected by mysql binary row based replication

#######


lib/mysql2_binlog_stream.rb
# encoding: UTF-8

Encoding.default_external = Encoding::UTF_8

###################################################################
#NOTE: include any stdlib first
###################################################################
require 'bigdecimal'
require 'open3'
require 'stringio'
require 'json'
require 'securerandom'
require 'digest'
require 'fcntl'
require 'fileutils'
require 'forwardable'
# ...


###################################################################
#NOTE: then bundler/gem dependencies next, after bundler/setup
require "rubygems"
require "bundler/setup"
require 'mysql2'
require 'superconfig2'
require 'json-diff'
# ...


#NOTE: this way we can ensure the proper libs are loaded
#NOTE: all other internalized depedencies autoloaded below
module Mysql2BinlogStream
  class Error < StandardError; end

  autoload 'Config', 'mysql2_binlog_stream/config'
  autoload 'Stream', 'mysql2_binlog_stream/stream'
  autoload 'Cli', 'mysql2_binlog_stream/cli'
  autoload 'App', 'mysql2_binlog_stream/app'
  autoload 'BinlogReader', 'mysql2_binlog_stream/binlog_reader'
  autoload 'BinlogFieldParser', 'mysql2_binlog_stream/binlog_field_parser'
  autoload 'BinlogEventParser', 'mysql2_binlog_stream/binlog_event_parser'
  autoload 'Binlog', 'mysql2_binlog_stream/binlog'
  autoload 'Observability', 'mysql2_binlog_stream/observability'
  autoload 'Cursor', 'mysql2_binlog_stream/cursor'
  autoload 'Fetcher', 'mysql2_binlog_stream/fetcher'

  #TODO: maybe build this at CD/CI time???
  #NOTE: this should be updated
  # A hash to map MySQL collation name to ID and character set name.
  #
  # This hash is produced by the following query:
  #   SELECT concat(
  #     "    :",
  #     rpad(collation_name, 24, " "),
  #     " => { :id => ",
  #     lpad(id, 3, " "),
  #     ", :character_set => :",
  #     rpad(character_set_name, 8, " "),
  #     " },"
  #   ) AS ruby_code
  #   FROM information_schema.collations
  #   ORDER BY collation_name
  #
  COLLATION_HASH = {
    :armscii8_bin             => { :id =>  64, :character_set => :armscii8 },
    :armscii8_general_ci      => { :id =>  32, :character_set => :armscii8 },
    :ascii_bin                => { :id =>  65, :character_set => :ascii    },
    :ascii_general_ci         => { :id =>  11, :character_set => :ascii    },
    :big5_bin                 => { :id =>  84, :character_set => :big5     },
    :big5_chinese_ci          => { :id =>   1, :character_set => :big5     },
    :binary                   => { :id =>  63, :character_set => :binary   },
    :cp1250_bin               => { :id =>  66, :character_set => :cp1250   },
    :cp1250_croatian_ci       => { :id =>  44, :character_set => :cp1250   },
    :cp1250_czech_cs          => { :id =>  34, :character_set => :cp1250   },
    :cp1250_general_ci        => { :id =>  26, :character_set => :cp1250   },
    :cp1250_polish_ci         => { :id =>  99, :character_set => :cp1250   },
    :cp1251_bin               => { :id =>  50, :character_set => :cp1251   },
    :cp1251_bulgarian_ci      => { :id =>  14, :character_set => :cp1251   },
    :cp1251_general_ci        => { :id =>  51, :character_set => :cp1251   },
    :cp1251_general_cs        => { :id =>  52, :character_set => :cp1251   },
    :cp1251_ukrainian_ci      => { :id =>  23, :character_set => :cp1251   },
    :cp1256_bin               => { :id =>  67, :character_set => :cp1256   },
    :cp1256_general_ci        => { :id =>  57, :character_set => :cp1256   },
    :cp1257_bin               => { :id =>  58, :character_set => :cp1257   },
    :cp1257_general_ci        => { :id =>  59, :character_set => :cp1257   },
    :cp1257_lithuanian_ci     => { :id =>  29, :character_set => :cp1257   },
    :cp850_bin                => { :id =>  80, :character_set => :cp850    },
    :cp850_general_ci         => { :id =>   4, :character_set => :cp850    },
    :cp852_bin                => { :id =>  81, :character_set => :cp852    },
    :cp852_general_ci         => { :id =>  40, :character_set => :cp852    },
    :cp866_bin                => { :id =>  68, :character_set => :cp866    },
    :cp866_general_ci         => { :id =>  36, :character_set => :cp866    },
    :cp932_bin                => { :id =>  96, :character_set => :cp932    },
    :cp932_japanese_ci        => { :id =>  95, :character_set => :cp932    },
    :dec8_bin                 => { :id =>  69, :character_set => :dec8     },
    :dec8_swedish_ci          => { :id =>   3, :character_set => :dec8     },
    :eucjpms_bin              => { :id =>  98, :character_set => :eucjpms  },
    :eucjpms_japanese_ci      => { :id =>  97, :character_set => :eucjpms  },
    :euckr_bin                => { :id =>  85, :character_set => :euckr    },
    :euckr_korean_ci          => { :id =>  19, :character_set => :euckr    },
    :gb2312_bin               => { :id =>  86, :character_set => :gb2312   },
    :gb2312_chinese_ci        => { :id =>  24, :character_set => :gb2312   },
    :gbk_bin                  => { :id =>  87, :character_set => :gbk      },
    :gbk_chinese_ci           => { :id =>  28, :character_set => :gbk      },
    :geostd8_bin              => { :id =>  93, :character_set => :geostd8  },
    :geostd8_general_ci       => { :id =>  92, :character_set => :geostd8  },
    :greek_bin                => { :id =>  70, :character_set => :greek    },
    :greek_general_ci         => { :id =>  25, :character_set => :greek    },
    :hebrew_bin               => { :id =>  71, :character_set => :hebrew   },
    :hebrew_general_ci        => { :id =>  16, :character_set => :hebrew   },
    :hp8_bin                  => { :id =>  72, :character_set => :hp8      },
    :hp8_english_ci           => { :id =>   6, :character_set => :hp8      },
    :keybcs2_bin              => { :id =>  73, :character_set => :keybcs2  },
    :keybcs2_general_ci       => { :id =>  37, :character_set => :keybcs2  },
    :koi8r_bin                => { :id =>  74, :character_set => :koi8r    },
    :koi8r_general_ci         => { :id =>   7, :character_set => :koi8r    },
    :koi8u_bin                => { :id =>  75, :character_set => :koi8u    },
    :koi8u_general_ci         => { :id =>  22, :character_set => :koi8u    },
    :latin1_bin               => { :id =>  47, :character_set => :latin1   },
    :latin1_danish_ci         => { :id =>  15, :character_set => :latin1   },
    :latin1_general_ci        => { :id =>  48, :character_set => :latin1   },
    :latin1_general_cs        => { :id =>  49, :character_set => :latin1   },
    :latin1_german1_ci        => { :id =>   5, :character_set => :latin1   },
    :latin1_german2_ci        => { :id =>  31, :character_set => :latin1   },
    :latin1_spanish_ci        => { :id =>  94, :character_set => :latin1   },
    :latin1_swedish_ci        => { :id =>   8, :character_set => :latin1   },
    :latin2_bin               => { :id =>  77, :character_set => :latin2   },
    :latin2_croatian_ci       => { :id =>  27, :character_set => :latin2   },
    :latin2_czech_cs          => { :id =>   2, :character_set => :latin2   },
    :latin2_general_ci        => { :id =>   9, :character_set => :latin2   },
    :latin2_hungarian_ci      => { :id =>  21, :character_set => :latin2   },
    :latin5_bin               => { :id =>  78, :character_set => :latin5   },
    :latin5_turkish_ci        => { :id =>  30, :character_set => :latin5   },
    :latin7_bin               => { :id =>  79, :character_set => :latin7   },
    :latin7_estonian_cs       => { :id =>  20, :character_set => :latin7   },
    :latin7_general_ci        => { :id =>  41, :character_set => :latin7   },
    :latin7_general_cs        => { :id =>  42, :character_set => :latin7   },
    :macce_bin                => { :id =>  43, :character_set => :macce    },
    :macce_general_ci         => { :id =>  38, :character_set => :macce    },
    :macroman_bin             => { :id =>  53, :character_set => :macroman },
    :macroman_general_ci      => { :id =>  39, :character_set => :macroman },
    :sjis_bin                 => { :id =>  88, :character_set => :sjis     },
    :sjis_japanese_ci         => { :id =>  13, :character_set => :sjis     },
    :swe7_bin                 => { :id =>  82, :character_set => :swe7     },
    :swe7_swedish_ci          => { :id =>  10, :character_set => :swe7     },
    :tis620_bin               => { :id =>  89, :character_set => :tis620   },
    :tis620_thai_ci           => { :id =>  18, :character_set => :tis620   },
    :ucs2_bin                 => { :id =>  90, :character_set => :ucs2     },
    :ucs2_czech_ci            => { :id => 138, :character_set => :ucs2     },
    :ucs2_danish_ci           => { :id => 139, :character_set => :ucs2     },
    :ucs2_esperanto_ci        => { :id => 145, :character_set => :ucs2     },
    :ucs2_estonian_ci         => { :id => 134, :character_set => :ucs2     },
    :ucs2_general_ci          => { :id =>  35, :character_set => :ucs2     },
    :ucs2_general_mysql500_ci => { :id => 159, :character_set => :ucs2     },
    :ucs2_hungarian_ci        => { :id => 146, :character_set => :ucs2     },
    :ucs2_icelandic_ci        => { :id => 129, :character_set => :ucs2     },
    :ucs2_latvian_ci          => { :id => 130, :character_set => :ucs2     },
    :ucs2_lithuanian_ci       => { :id => 140, :character_set => :ucs2     },
    :ucs2_persian_ci          => { :id => 144, :character_set => :ucs2     },
    :ucs2_polish_ci           => { :id => 133, :character_set => :ucs2     },
    :ucs2_romanian_ci         => { :id => 131, :character_set => :ucs2     },
    :ucs2_roman_ci            => { :id => 143, :character_set => :ucs2     },
    :ucs2_sinhala_ci          => { :id => 147, :character_set => :ucs2     },
    :ucs2_slovak_ci           => { :id => 141, :character_set => :ucs2     },
    :ucs2_slovenian_ci        => { :id => 132, :character_set => :ucs2     },
    :ucs2_spanish2_ci         => { :id => 142, :character_set => :ucs2     },
    :ucs2_spanish_ci          => { :id => 135, :character_set => :ucs2     },
    :ucs2_swedish_ci          => { :id => 136, :character_set => :ucs2     },
    :ucs2_turkish_ci          => { :id => 137, :character_set => :ucs2     },
    :ucs2_unicode_ci          => { :id => 128, :character_set => :ucs2     },
    :ujis_bin                 => { :id =>  91, :character_set => :ujis     },
    :ujis_japanese_ci         => { :id =>  12, :character_set => :ujis     },
    :utf16_bin                => { :id =>  55, :character_set => :utf16    },
    :utf16_czech_ci           => { :id => 111, :character_set => :utf16    },
    :utf16_danish_ci          => { :id => 112, :character_set => :utf16    },
    :utf16_esperanto_ci       => { :id => 118, :character_set => :utf16    },
    :utf16_estonian_ci        => { :id => 107, :character_set => :utf16    },
    :utf16_general_ci         => { :id =>  54, :character_set => :utf16    },
    :utf16_hungarian_ci       => { :id => 119, :character_set => :utf16    },
    :utf16_icelandic_ci       => { :id => 102, :character_set => :utf16    },
    :utf16_latvian_ci         => { :id => 103, :character_set => :utf16    },
    :utf16_lithuanian_ci      => { :id => 113, :character_set => :utf16    },
    :utf16_persian_ci         => { :id => 117, :character_set => :utf16    },
    :utf16_polish_ci          => { :id => 106, :character_set => :utf16    },
    :utf16_romanian_ci        => { :id => 104, :character_set => :utf16    },
    :utf16_roman_ci           => { :id => 116, :character_set => :utf16    },
    :utf16_sinhala_ci         => { :id => 120, :character_set => :utf16    },
    :utf16_slovak_ci          => { :id => 114, :character_set => :utf16    },
    :utf16_slovenian_ci       => { :id => 105, :character_set => :utf16    },
    :utf16_spanish2_ci        => { :id => 115, :character_set => :utf16    },
    :utf16_spanish_ci         => { :id => 108, :character_set => :utf16    },
    :utf16_swedish_ci         => { :id => 109, :character_set => :utf16    },
    :utf16_turkish_ci         => { :id => 110, :character_set => :utf16    },
    :utf16_unicode_ci         => { :id => 101, :character_set => :utf16    },
    :utf32_bin                => { :id =>  61, :character_set => :utf32    },
    :utf32_czech_ci           => { :id => 170, :character_set => :utf32    },
    :utf32_danish_ci          => { :id => 171, :character_set => :utf32    },
    :utf32_esperanto_ci       => { :id => 177, :character_set => :utf32    },
    :utf32_estonian_ci        => { :id => 166, :character_set => :utf32    },
    :utf32_general_ci         => { :id =>  60, :character_set => :utf32    },
    :utf32_hungarian_ci       => { :id => 178, :character_set => :utf32    },
    :utf32_icelandic_ci       => { :id => 161, :character_set => :utf32    },
    :utf32_latvian_ci         => { :id => 162, :character_set => :utf32    },
    :utf32_lithuanian_ci      => { :id => 172, :character_set => :utf32    },
    :utf32_persian_ci         => { :id => 176, :character_set => :utf32    },
    :utf32_polish_ci          => { :id => 165, :character_set => :utf32    },
    :utf32_romanian_ci        => { :id => 163, :character_set => :utf32    },
    :utf32_roman_ci           => { :id => 175, :character_set => :utf32    },
    :utf32_sinhala_ci         => { :id => 179, :character_set => :utf32    },
    :utf32_slovak_ci          => { :id => 173, :character_set => :utf32    },
    :utf32_slovenian_ci       => { :id => 164, :character_set => :utf32    },
    :utf32_spanish2_ci        => { :id => 174, :character_set => :utf32    },
    :utf32_spanish_ci         => { :id => 167, :character_set => :utf32    },
    :utf32_swedish_ci         => { :id => 168, :character_set => :utf32    },
    :utf32_turkish_ci         => { :id => 169, :character_set => :utf32    },
    :utf32_unicode_ci         => { :id => 160, :character_set => :utf32    },
    :utf8mb4_bin              => { :id =>  46, :character_set => :utf8mb4  },
    :utf8mb4_czech_ci         => { :id => 234, :character_set => :utf8mb4  },
    :utf8mb4_danish_ci        => { :id => 235, :character_set => :utf8mb4  },
    :utf8mb4_esperanto_ci     => { :id => 241, :character_set => :utf8mb4  },
    :utf8mb4_estonian_ci      => { :id => 230, :character_set => :utf8mb4  },
    :utf8mb4_general_ci       => { :id =>  45, :character_set => :utf8mb4  },
    :utf8mb4_hungarian_ci     => { :id => 242, :character_set => :utf8mb4  },
    :utf8mb4_icelandic_ci     => { :id => 225, :character_set => :utf8mb4  },
    :utf8mb4_latvian_ci       => { :id => 226, :character_set => :utf8mb4  },
    :utf8mb4_lithuanian_ci    => { :id => 236, :character_set => :utf8mb4  },
    :utf8mb4_persian_ci       => { :id => 240, :character_set => :utf8mb4  },
    :utf8mb4_polish_ci        => { :id => 229, :character_set => :utf8mb4  },
    :utf8mb4_romanian_ci      => { :id => 227, :character_set => :utf8mb4  },
    :utf8mb4_roman_ci         => { :id => 239, :character_set => :utf8mb4  },
    :utf8mb4_sinhala_ci       => { :id => 243, :character_set => :utf8mb4  },
    :utf8mb4_slovak_ci        => { :id => 237, :character_set => :utf8mb4  },
    :utf8mb4_slovenian_ci     => { :id => 228, :character_set => :utf8mb4  },
    :utf8mb4_spanish2_ci      => { :id => 238, :character_set => :utf8mb4  },
    :utf8mb4_spanish_ci       => { :id => 231, :character_set => :utf8mb4  },
    :utf8mb4_swedish_ci       => { :id => 232, :character_set => :utf8mb4  },
    :utf8mb4_turkish_ci       => { :id => 233, :character_set => :utf8mb4  },
    :utf8mb4_unicode_ci       => { :id => 224, :character_set => :utf8mb4  },
    :utf8_bin                 => { :id =>  83, :character_set => :utf8     },
    :utf8_czech_ci            => { :id => 202, :character_set => :utf8     },
    :utf8_danish_ci           => { :id => 203, :character_set => :utf8     },
    :utf8_esperanto_ci        => { :id => 209, :character_set => :utf8     },
    :utf8_estonian_ci         => { :id => 198, :character_set => :utf8     },
    :utf8_general_ci          => { :id =>  33, :character_set => :utf8     },
    :utf8_general_mysql500_ci => { :id => 223, :character_set => :utf8     },
    :utf8_hungarian_ci        => { :id => 210, :character_set => :utf8     },
    :utf8_icelandic_ci        => { :id => 193, :character_set => :utf8     },
    :utf8_latvian_ci          => { :id => 194, :character_set => :utf8     },
    :utf8_lithuanian_ci       => { :id => 204, :character_set => :utf8     },
    :utf8_persian_ci          => { :id => 208, :character_set => :utf8     },
    :utf8_polish_ci           => { :id => 197, :character_set => :utf8     },
    :utf8_romanian_ci         => { :id => 195, :character_set => :utf8     },
    :utf8_roman_ci            => { :id => 207, :character_set => :utf8     },
    :utf8_sinhala_ci          => { :id => 211, :character_set => :utf8     },
    :utf8_slovak_ci           => { :id => 205, :character_set => :utf8     },
    :utf8_slovenian_ci        => { :id => 196, :character_set => :utf8     },
    :utf8_spanish2_ci         => { :id => 206, :character_set => :utf8     },
    :utf8_spanish_ci          => { :id => 199, :character_set => :utf8     },
    :utf8_swedish_ci          => { :id => 200, :character_set => :utf8     },
    :utf8_turkish_ci          => { :id => 201, :character_set => :utf8     },
    :utf8_unicode_ci          => { :id => 192, :character_set => :utf8     },
  }

  # An array of collation IDs to collation and character set name for
  # efficient lookup by ID.
  COLLATION = COLLATION_HASH.inject(Array.new) do |collation_array, item|
    collation_array[item[1][:id]] = { 
      :character_set => item[1][:character_set],
      :collation     => item[0],
    }
    collation_array
  end

  #TODO: improve SUPERCONFIG/ENV interface!!?
  SUPERCONFIG = Config.new
end
